GHI #13 Setup tests

doodspav opened the pull request:
## Scope

- create core test files (`test`)
- setup CMake to build and install tests
- each test has its own target as well as targets for test kind, and target for all tests
- each test kind group is its own component and can be installed per component
- all tests can also be installed together
- path is automatically set for tests run with ctest on windows (unless disabled via option)
- file containing path for test dependencies can be created on windows if option is enabled
- tests can be built as a completely independent project
- user can choose which tests to build using regex option

doodspav committed:
GHI #13 Add patomic_BUILD_TESTING override option

Signed-off-by: doodspav <doodspav@gmail.com>

doodspav committed:
GHI #13 Implement minimal test with CMake support

Signed-off-by: doodspav <doodspav@gmail.com>

doodspav committed:
GHI #13 Fix Windows PATH issues for tests

Signed-off-by: doodspav <doodspav@gmail.com>

doodspav committed:
GHI #13 Minor comment change

Signed-off-by: doodspav <doodspav@gmail.com>

doodspav committed:
GHI #13 Enable installing tests

Signed-off-by: doodspav <doodspav@gmail.com>

doodspav committed:
GHI #13 Prevent in-source build of tests

Signed-off-by: doodspav <doodspav@gmail.com>

doodspav committed:
GHI #13 Support install tests from multiple directories

Signed-off-by: doodspav <doodspav@gmail.com>

doodspav committed:
GHI #13 Refactor test CMake to do Windows stuff per-component

- windows dependency file is generated per test kind, rather than once for all tests
- all windows stuff is shifted either into test kind CMake file, or into create_test
- create_test now accepts LINK parameters
- create_test now uses the "name" parameter as the created target name

Signed-off-by: doodspav <doodspav@gmail.com>

doodspav committed:
GHI #13 Forgot PARENT_SCOPE in windows_deps_path

- also tests should have unique suite.name or CTest will skip "duplicates"

Signed-off-by: doodspav <doodspav@gmail.com>

doodspav committed:
GHI #13 Abstract functionality into create_test_win_deps_path_file

Signed-off-by: doodspav <doodspav@gmail.com>

doodspav committed:
GHI #13 Update target and executable created by create_test

create_test(BT example SOURCE example.cpp):
- target created -> patomic_bt_example
- executable created -> example(.exe)
- executable installed as -> <prefix>/patomic/bt/example(.exe)

Signed-off-by: doodspav <doodspav@gmail.com>

doodspav committed:
GHI #13 Add patomic_test_CREATE_TEST_TARGETS_MATCHING option

- only creates test targets that match the regex obtained from this option
- the target name must match the regex; it is not tested against:
  - executable name
  - name passed to create_test
  - test suite/case name (this can be limited at runtime with gtest_filter)
- look at create_test docs to see what target name is chosen

Signed-off-by: doodspav <doodspav@gmail.com>

doodspav committed:
GHI #13 Add patomic_test component that installs all tests

- also remove redundant example_add.cpp file

Signed-off-by: doodspav <doodspav@gmail.com>

doodspav committed:
GHI #13 Add debug messages when skipping creation of something

- when skipping creation of test due to not matching regex
- when skipping creation of deps_path file due to no tests existing

Signed-off-by: doodspav <doodspav@gmail.com>

GabrielRavier reviewed (dismissed):
LGTM

GabrielRavier commented on CMakeLists.txt:
```diff
@@ -79,3 +79,11 @@ target_compile_features(patomic_patomic PUBLIC c_std_90)
+# ---- Setup Tests ----
+
+include(CTest)
+if(patomic_BUILD_TESTING)
```
note: when I was reading this I wrongly assumed I was supposed to use `patomic_BUILD_TESTING` to enable testing (instead of `BUILD_TESTING`) and this somehow screwed up ctest, making it think there were no tests at all - looking at the rest of the code doesn't seem to indicate any way in which not setting `BUILD_TESTING` could cause any problems, so I'm left with the assumption that `ctest` must be directly checking for `BUILD_TESTING` or something like that.

doodspav replied:
Fixed.

doodspav committed:
GHI #13 Ensure tests still build if BUILD_TESTING is disabled but overriden

- if BUILD_TESTING=OFF and patomic_BUILD_TESTING=ON, tests are built but not registered with CTest
- need to call enable_testing() to get CTest to behave as expected

Signed-off-by: doodspav <doodspav@gmail.com>

doodspav committed:
GHI #13 Add `patomic_test_CREATE_WINDOWS_PATH_FILE` option

- disabled by default
- determines whether windows_dependencies_path.txt is generated for tests or not

Signed-off-by: doodspav <doodspav@gmail.com>

GabrielRavier reviewed (approved):
LGTM
